src/core/DY.Core.Trace.Base.fst

```diff
@@ -409,6 +509,16 @@ let rand_generated_at #label_t tr i b =
     time == i /\ (exists usg lab. entry_at tr i (RandGen usg lab len))
   | _ -> False
 
+
+/// Was a random bytestring generated just now,
+/// i.e., at the end of the trace?
+
+val rand_just_generated: #label_t:Type -> trace_ label_t -> bytes -> prop
+let rand_just_generated tr rand =
+  1 <= trace_length tr /\ // or `is_not_empty tr` which I prefer but requires fuel 1 on the protocol level
```

Comment:
Ah, this answers the question about raising `fuel` in NSL: this is an artifact of a previous version that used `is_not_empty` here.

I think we should add a theorem like
```fstar
val trace_length_is_not_empty:
  #label_t:Type ->
  tr:trace_ label_t ->
  Lemma
  (is_not_empty tr <==> 1 <= trace_length tr)
  [SMTPat (is_not_empty tr)]
let trace_length_is_not_empty #label_t tr = () // probably needs a fuel / ifuel of 1
```
That would probably help with this https://github.com/REPROSEC/dolev-yao-star-extrinsic/issues/84#issuecomment-2493970063

Reply:
Indeed, as I mentioned in the description of this PR, the increased fuel in the NSL example is related to using `is_not_empty` here.
Again, this comment is meant as discussion:
Do people prefer sticking to `trace_length`?
I prefer not to. And yes, if we go for `is_not_empty`, I would add a lemma like you suggested.

Reply:
Both are fine for me, but in any case we need the lemma relating `is_not_empty tr` and `1 <= trace_length tr` with an SMT pattern on `is_not_empty tr`.

Reply:
i added a corresponding lemma.
with the smt pattern, we don't need the extra fuel in the NSL example anymore.